### PR TITLE
Adding subsection permissions for Authentication

### PR DESCRIPTION
**Pull request by BenCookie95**
#### Summary
Adding subsection permissions for Authentication

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-32349

### CODE REVIEW

File: components/admin_console/admin_definition.jsx [hunk 1]
```diff
@@ -3825,8 +3857,11 @@ const AdminDefinition = {
                         label: t('admin.saml.enableAdminAttrTitle'),
                         label_default: 'Enable Admin Attribute:',
                         isDisabled: it.any(
+
+                            // Should this be removed?
                             it.not(it.isSystemAdmin),
```
Comment (hahmadia) [inline components/admin_console/admin_definition.jsx:3862]: No this attribute can make people a system admin allowing them to manipulate more than what they have control too. Leave it in.

File: components/admin_console/admin_definition.jsx [hunk 2]
```diff
@@ -3840,9 +3875,12 @@ const AdminDefinition = {
                         help_text_default: '(Optional) The attribute in the SAML Assertion for designating System Admins. The users selected by the query will have access to your Mattermost server as System Admins. By default, System Admins have complete access to the Mattermost System Console.\n \nExisting members that are identified by this attribute will be promoted from member to System Admin upon next login. The next login is based upon Session lengths set in **System Console > Session Lengths.** It is highly recommend to manually demote users to members in **System Console > User Management** to ensure access is restricted immediately.\n \nNote: If this filter is removed/changed, System Admins that were promoted via this filter will be demoted to members and will not retain access to the System Console. When this filter is not in use, System Admins can be manually promoted/demoted in **System Console > User Management**.',
                         help_text_markdown: true,
                         isDisabled: it.any(
+
+                            // Should this be removed?
                             it.not(it.isSystemAdmin),
```
Comment (hahmadia) [inline components/admin_console/admin_definition.jsx:3880]: No this attribute can make people a system admin allowing them to manipulate more than what they have control too. Leave it in.

File: e2e/cypress/support/api/role.js [hunk 3]
```diff
@@ -284,7 +317,14 @@ export const defaultRolesPermissions = {
...
         'delete_public_channel',
-        'sysconsole_read_authentication',
+        'sysconsole_read_authentication_signup',
+        'sysconsole_read_authentication_email',
+        'sysconsole_read_authentication_password',
+        'sysconsole_read_authentication_mfa',
+        'sysconsole_read_authentication_ldap',
+        'sysconsole_read_authentication_saml',
+        'sysconsole_read_authentication_openid',
+        'sysconsole_read_authentication_guest_access',
```
Comment (hahmadia) [inline e2e/cypress/support/api/role.js:327]: Please also add any ancillary permissions that come with these.

File: e2e/cypress/support/api/role.js [hunk 4]
```diff
@@ -237,7 +263,14 @@ export const defaultRolesPermissions = {
...
         'sysconsole_read_about',
-        'sysconsole_read_authentication',
+        'sysconsole_read_authentication_signup',
+        'sysconsole_read_authentication_email',
+        'sysconsole_read_authentication_password',
+        'sysconsole_read_authentication_mfa',
+        'sysconsole_read_authentication_ldap',
+        'sysconsole_read_authentication_saml',
+        'sysconsole_read_authentication_openid',
+        'sysconsole_read_authentication_guest_access',
```
Comment (hahmadia) [inline e2e/cypress/support/api/role.js:273]: Please also add any ancillary permissions that come with these.

File: e2e/cypress/support/api/role.js [hunk 5]
```diff
@@ -187,7 +206,14 @@ export const defaultRolesPermissions = {
...
         'list_private_teams',
-        'sysconsole_read_authentication',
+        'sysconsole_read_authentication_signup',
+        'sysconsole_read_authentication_email',
+        'sysconsole_read_authentication_password',
+        'sysconsole_read_authentication_mfa',
+        'sysconsole_read_authentication_ldap',
+        'sysconsole_read_authentication_saml',
+        'sysconsole_read_authentication_openid',
+        'sysconsole_read_authentication_guest_access',
```
Comment (hahmadia) [inline e2e/cypress/support/api/role.js:216]: Please also add any ancillary permissions that come with these.

**Review by hahmadia — APPROVED** [review]
LGTM! Minor changes but great stuff.
Going to be adding Scott as a reviewer since this touches on the OpenID and I think it be nice to get his eyes on it too.

File: components/admin_console/admin_definition.jsx [hunk 6]
```diff
@@ -2961,7 +2978,10 @@ const AdminDefinition = {
                         label: t('admin.ldap.enableAdminFilterTitle'),
                         label_default: 'Enable Admin Filter:',
                         isDisabled: it.any(
+
+                            // Should this one be removed?
                             it.not(it.isSystemAdmin),
+                            it.not(it.userHasWritePermissionOnResource(RESOURCE_KEYS.AUTHENTICATION.LDAP)),
```
Comment (sbishel) [inline components/admin_console/admin_definition.jsx:2984]: No, only a "SystemAdmin" can update permissions which changes "SystemAdmins".  Therefore this check should remain. No additional check is necessary.

File: components/admin_console/admin_definition.jsx [hunk 7]
```diff
@@ -2979,6 +2999,9 @@ const AdminDefinition = {
                         placeholder: t('admin.ldap.adminFilterEx'),
                         placeholder_default: 'E.g.: "(objectClass=user)"',
                         isDisabled: it.any(
+                            it.not(it.userHasWritePermissionOnResource(RESOURCE_KEYS.AUTHENTICATION.LDAP)),
+
+                            // Should this one be removed?
```
Comment (sbishel) [inline components/admin_console/admin_definition.jsx:3004]: No, see above.

File: components/admin_console/admin_definition.jsx [hunk 8]
```diff
@@ -3428,6 +3452,7 @@ const AdminDefinition = {
                         type: Constants.SettingsTypes.TYPE_CUSTOM,
                         component: LDAPFeatureDiscovery,
                         key: 'LDAPFeatureDiscovery',
+                        isDisabled: it.not(it.userHasWritePermissionOnResource(RESOURCE_KEYS.AUTHENTICATION.LDAP)),
```
Comment (sbishel) [inline components/admin_console/admin_definition.jsx:3455]: This allows a user to request a trial license. Therefore, the user requires "manage_license_information" permission which is tied to ABOUT.EDITION_AND_LICENSE. 
The ABOUT PRs may need merged first.

```suggestion
                        isDisabled: it.not(it.userHasWritePermissionOnResource(RESOURCE_KEYS.ABOUT.EDITION_AND_LICENSE)),
```

File: components/admin_console/admin_definition.jsx [hunk 9]
```diff
@@ -3417,6 +3440,7 @@ const AdminDefinition = {
             title_default: 'AD/LDAP',
             isHidden: it.any(
                 it.licensedForFeature('LDAP'),
+                it.not(it.userHasReadPermissionOnResource(RESOURCE_KEYS.AUTHENTICATION.LDAP)),
```
Comment (sbishel) [inline components/admin_console/admin_definition.jsx:3443]: I don't think we want a check on this resource. I believe everyone should be able to see it.

File: components/admin_console/admin_definition.jsx [hunk 10]
```diff
@@ -3938,6 +3976,7 @@ const AdminDefinition = {
             title: t('admin.sidebar.saml'),
             title_default: 'SAML 2.0',
             isHidden: it.any(
+                it.not(it.userHasReadPermissionOnResource(RESOURCE_KEYS.AUTHENTICATION.SAML)),
```
Comment (sbishel) [inline components/admin_console/admin_definition.jsx:3979]: Again feature discovery, no permission necessary.

File: components/admin_console/admin_definition.jsx [hunk 11]
```diff
@@ -3950,6 +3989,7 @@ const AdminDefinition = {
                         type: Constants.SettingsTypes.TYPE_CUSTOM,
                         component: SAMLFeatureDiscovery,
                         key: 'SAMLFeatureDiscovery',
+                        isDisabled: it.not(it.userHasWritePermissionOnResource(RESOURCE_KEYS.AUTHENTICATION.SAML)),
```
Comment (sbishel) [inline components/admin_console/admin_definition.jsx:3992]: Allows to create a trial license. Will require write permisson on ABOUT_ EDITION_AND_LICENSE

File: e2e/cypress/support/api/role.js [hunk 12]
```diff
@@ -74,7 +81,15 @@ export const defaultRolesPermissions = {
         'manage_system_wide_oauth',
         'manage_private_channel_members',
         'revoke_user_access_token',
-        'sysconsole_read_authentication',
+        'sysconsole_write_user_management_channels',
```
Comment (sbishel) [inline e2e/cypress/support/api/role.js:84]: adding twice, also on line 61

File: plugins/test/__snapshots__/main_menu_action.test.jsx.snap [hunk 13]
```diff
@@ -305,10 +305,20 @@ exports[`plugins/MainMenuActions should match snapshot and click plugin item for
...
+          "sysconsole_read_authentication_saml",
+          "sysconsole_read_authentication_openid",
+          "sysconsole_read_authentication_guest_access",
           "sysconsole_read_plugins",
           "sysconsole_read_integrations",
-          "sysconsole_read_compliance",
+          "sysconsole_read_compliance_data_retention_policy",
+          "sysconsole_read_compliance_compliance_export",
+          "sysconsole_read_compliance_compliance_monitoring",
+          "sysconsole_read_compliance_custom_terms_of_service",
```
Comment (sbishel) [inline plugins/test/__snapshots__/main_menu_action.test.jsx.snap:321]: I added these with "integrations" as well. Just FYI in case of merge conflicts.

**Review by sbishel — CHANGES_REQUESTED** [review]

File: components/admin_console/admin_definition.jsx [hunk 14]
```diff
@@ -3270,7 +3293,7 @@ const AdminDefinition = {
                         success_message: t('admin.ldap.testSuccess'),
                         success_message_default: 'AD/LDAP Test Successful',
                         isDisabled: it.any(
-                            it.not(it.userHasReadPermissionOnResource('authentication')),
+                            it.not(it.userHasReadPermissionOnResource(RESOURCE_KEYS.AUTHENTICATION.LDAP)),
```
Comment (hahmadia) [inline components/admin_console/admin_definition.jsx:3296]: This should be a Write Permission and not a read as Max pointed out in the Server PR.

**Review by hahmadia — CHANGES_REQUESTED** [review]
One minor change

File: components/admin_console/admin_definition.jsx (hunk 14 above)
Comment (hahmadia) [reply components/admin_console/admin_definition.jsx:3296]: Give the ancillary of this endpoint to WRITE_AUTH instead of READ_AUTH on server side

**Review by hahmadia — APPROVED** [review]

**Review by maxerenberg — APPROVED** [review]

**Review by sbishel — APPROVED** [review]
@BenCookie95, @hahmadia Please review the last commit. Fixes a couple issues - 
- When only OpenId has access, section was hidden.
- if access to any authentication section, but not OpenId, threw javascript error

**Comment by hahmadia** [comment]
Looks good @sbishel . Seems a test just needs to be fixed.

**Pull request merged**